### PR TITLE
Doxygen cpp

### PR DESCRIPTION
**Pull request by mislavn**
Add doxygen documentation for C++ parts which reference libyang C API.

@rkrejci can you take a look at the changes and the generated graphs, to generate graphs just install `graphviz`, cmake will do the rest.

There are a few things I want to discuss:
1) I added *.cpp files to Doxygen input, so the source is referenced. Do you want me to remove it?
2) The macros are expanded because when you print a Python object, it will show the expanded C++ type. Is this an issue for the C documentation?
3) Should I disable generating C++ documentation if `GEN_LANGUAGE_BINDINGS` is set to `NO`?

### CODE REVIEW

File: swig/cpp/src/Tree_Schema.hpp [hunk 1]
```diff
@@ -530,18 +716,24 @@ class Schema_Node_Leaf : public Schema_Node
...
     S_Set backlinks();
+    /** get when variable from [lys_node_leaf](@ref lys_node_leaf)*/
     S_When when();
+    /** get type variable from [lys_node_leaf](@ref lys_node_leaf)*/
     S_Type type();
+    /** get units variable from [lys_node_leaf](@ref lys_node_leaf)*/
     const char *units() {return ((struct lys_node_leaf *)node)->units;};
+    /** get dflt variable from [lys_node_leaf](@ref lys_node_leaf)*/
     const char *dflt() {return ((struct lys_node_leaf *)node)->dflt;};
-    S_Schema_Node child() override {return nullptr;};
```
[the remark this answers is not in the document]
Comment (mislavn) [reply swig/cpp/src/Tree_Schema.hpp]: No, I failed to refactor it and forgot to return the line.

File: CMakeLists.txt [hunk 2]
```diff
@@ -300,7 +300,13 @@ endif()
 # generate doxygen documentation for libyang API
 find_package(Doxygen)
 if(DOXYGEN_FOUND)
-    set(DOXYGEN_SKIP_DOT TRUE)
+    find_program(DOT_PATH dot PATH_SUFFIXES graphviz2.38/bin graphviz/bin)
+    if(DOT_PATH)
+        set(HAVE_DOT "YES")
+    else()
+        set(HAVE_DOT "NO")
+        message(AUTHOR_WARNING "Doxygen: to genaret UML diagrams please install graphviz")
```
[the remark this answers is not in the document]
Comment (mislavn) [reply CMakeLists.txt]: Done.

File: swig/cpp/src/Tree_Schema.hpp [hunk 3]
```diff
@@ -476,15 +654,20 @@ class Schema_Node_Container : public Schema_Node
             throw std::invalid_argument("Type must be LYS_CONTAINER");
         }
     };
+    /** wrapper for struct [lys_node_container](@ref lys_node_container), for internal use only */
```
[the remark this answers is not in the document]
Comment (mislavn) [reply swig/cpp/src/Tree_Schema.hpp]: All of the functions and class methods which have libyang C structs are for internal use (there are some exceptions for libnetconf2 Python bindings), but there should be a way of getting the same functionality.

Do you mean something like this?
```
        for(auto elem = schema_list->begin() ; elem != schema_list->end() ; ++elem) {
            std::cout << "schema name " << (*elem)->name() << " type " << (*elem)->nodetype() << std::endl;
            if  (LYS_LEAF == (*elem)->nodetype()) {
                auto leaf = Schema_Node_Leaf(*elem);
```

[the remark this answers is not in the document]
Comment (jktjkt) [reply swig/cpp/src/Tree_Schema.hpp]: Right, sorry, I missed the point that there are two constructors here.

**Comment by mislavn** [comment]
https://github.com/CESNET/libyang/pull/565/commits/679063c899b1cf00d6f468a00e2ffe2639344584 fixes https://github.com/CESNET/libyang/issues/564.

**Comment by mislavn** [comment]
Actually, still need to make some fixes for the list/vector support in the Python bindings.

**Comment by mislavn** [comment]
@rkrejci con you take a look at the generated documentation and graphs?

To generate graphs you need graphviz `sudo apt-get install graphviz`.

File: swig/cpp/examples/process_tree.cpp [hunk 4]
```diff
@@ -75,6 +75,13 @@ int main() {
         auto schema_list = std::shared_ptr<std::vector<S_Schema_Node>>(node->schema()->tree_dfs());
         for(auto elem = schema_list->begin() ; elem != schema_list->end() ; ++elem) {
             std::cout << "schema name " << (*elem)->name() << " type " << (*elem)->nodetype() << std::endl;
+            if  (LYS_LEAF == (*elem)->nodetype()) {
```
Comment (jktjkt) [inline swig/cpp/examples/process_tree.cpp]: This is not related to Doxygen, and therefore it should be made in a separate commit.

File: swig/cpp/src/Tree_Data.cpp [hunk 5]
```diff
@@ -208,6 +208,17 @@ S_Data_Node Data_Node::dup(int recursive) {
     S_Deleter new_deleter = std::make_shared<Deleter>(new_node, deleter);
     return std::make_shared<Data_Node>(new_node, new_deleter);
 }
+S_Data_Node Data_Node::dup_withsiblings(int recursive) {
```
Comment (jktjkt) [inline swig/cpp/src/Tree_Data.cpp]: not related to Doxygen

File: swig/cpp/src/Tree_Schema.cpp [hunk 6]
```diff
@@ -335,14 +335,11 @@ Schema_Node_Leaf::~Schema_Node_Leaf() {};
 S_Set Schema_Node_Leaf::backlinks() LY_NEW_CASTED(lys_node_leaf, node, backlinks, Set);
 S_When Schema_Node_Leaf::when() LY_NEW_CASTED(lys_node_leaf, node, when, When);
 S_Type Schema_Node_Leaf::type() {return std::make_shared<Type>(&((struct lys_node_leaf *)node)->type, deleter);}
-int Schema_Node_Leaf::is_key() {
+S_Schema_Node_List Schema_Node_Leaf::is_key() {
     uint8_t pos;
 
-    if (lys_is_key((struct lys_node_leaf *)node, &pos)) {
```
Comment (jktjkt) [inline swig/cpp/src/Tree_Schema.cpp]: not a doxygen change

File: swig/swig_base/cpp_classes.i [hunk 7]
```diff
@@ -259,7 +261,7 @@
 %newobject Schema_Node_Leaflist::xpath_atomize;
 %ignore    Schema_Node_Leaflist::swig_node;
 %ignore    Schema_Node_Leaflist::swig_deleter;
-%newobject Schema_Node_Leaf::type;
```
Comment (jktjkt) [inline swig/swig_base/cpp_classes.i]: not a doxygen change

**Comment by mislavn** [comment]
While writing the documentation found some bugs and immediately made bug fixes but kept them in the same branch for convenience. They are now removed.
